prometheus/example_timer_test.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2014 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus_test
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	requestDuration = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:    "example_request_duration_seconds",
+		Help:    "Histogram for the runtime of a simple example function.",
+		Buckets: prometheus.LinearBuckets(0.01, 0.01, 10),
+	})
+)
+
+func ExampleTimer() {
+	// timer times this example function. It uses a Histogram, but a Summary
+	// would also work, as both implement Observer. Check out
+	// https://prometheus.io/docs/practices/histograms/ for differences.
+	timer := prometheus.NewTimer(requestDuration)
```

Comment:
You _can_ write it as a one-liner:
```Go
defer prometheus.NewTimer(requestDuration).ObserveDuration()
```

This kind of timer pattern is pretty common in Go, and what I have seen multiple times in Go conference talks is the recommendation to write it in two lines for clarity. That's why I did the examples that way. You can still write it in one line if you like.

Another possible simplification is to return a naked function to call for the observation:
```Go
observeDuration := prometheus.NewTimer(requestDuration)
defer observeDuration()
```

Or in one line (quite obfuscated):
```Go
defer prometheus.NewTimer(requestDuration)()
```

To avoid the latter way, you can just tell people to give the returned function a nicely named variable as in the 2nd to last example.

I decided to go the way with an explicit method because it is in better agreement with the client library and makes the one-liner better to read. My impression is that this is in the sweet spot where we are in maximum agreement with most of the community and the Prometheus client library guidelines.